src/atldld/sync.py

```diff
@@ -260,59 +227,190 @@ def download_parallel_dataset(
     downsample_img
         The downloaded image will have both the height and the width
         downsampled by `2 ** downsample_img`.
-
-    Returns
-    -------
-    res_dict : generator
-        Generator yielding consecutive four tuples of
-        (image_id, constant_ref_coordinate, img, df).
-        The `constant_ref_coordinate` is the dimension in the given axis in microns.
-        The `img` is the raw gene expression image with dtype `uint8`.
-        The `df` is the displacement field.
-        Note that the sorting. If `include_expression=True` then last returned image
-        is the processed expression image.
-        That is the generator yield (image_id, p, img, df, img_expr).
     """
-    metadata_2d_dict = get_2d_bulk(
-        dataset_id,
-        ref2inp=True,
-    )
-    metadata_2d = sorted(
-        [
-            (image_id, affine_2d, section_number)
-            for image_id, (affine_2d, section_number) in metadata_2d_dict.items()
-        ],
-        key=lambda x: -int(x[2]),  # we use section_number for sorting
-    )
-
-    affine_3d = get_3d(
-        dataset_id,
-        ref2inp=True,
-        return_meta=False,
-    )
-    axis = CommonQueries.get_axis(dataset_id)
-
-    for image_id, affine_2d, _ in metadata_2d:
-        p, i, r = xy_to_pir_API_single(*detection_xy, image_id=image_id)
-        slice_ref_coordinate = p if axis == "coronal" else r
 
-        df = get_parallel_transform(
-            slice_ref_coordinate,
-            affine_2d,
-            affine_3d,
-            downsample_ref=downsample_ref,
-            axis=axis,
-            downsample_img=downsample_img,
+    def __init__(
+        self,
+        dataset_id: int,
+        downsample_ref: int = 25,
+        detection_xy: Tuple[float, float] = (0, 0),
+        include_expression: bool = False,
+        downsample_img: int = 0,
+    ):
+        self.dataset_id = dataset_id
+        self.downsample_ref = downsample_ref
+        self.detection_xy = detection_xy
+        self.include_expression = include_expression
+        self.downsample_img = downsample_img
+
+        self.metadata: Dict[str, Any] = {}
+        # populated by calling `fetch_metadata`
+
+    def __len__(self) -> int:
+        """Return the number of images in the dataset."""
+        if "images" not in self.metadata:
+            self.fetch_metadata()
```

Comment:
Heh, I would raise an error instead and only fetch the metadata when the user explicitly calls `fetch_metadata()`

Anyway, it is a detail and a matter of taste I guess. Your call.

Reply:
Done in feb3b63.